### DISCUSSION TITLE
Fails to refresh access token

### DISCUSSION THREAD

**Discussion by HonestJohn61** [mikejac/node-red-contrib-google-smarthome#531] [General]
About 2 or 3 times a month I have to restart my Node-RED flows when the link fails. Here's the Google Actions Console error log for when this happened just now: 

{"@type":"type.googleapis.com/google.identity.accountlinking.type.AccountLinkingError", "errorReason":"Failed to get response from 3P. 3P returned 5XX", "request":{…}, "response":{…}, "sessionId":-1.460215712E9, "step":"REFRESH_ACCESS_TOKEN"}

My server restarts my Node-RED instance every 24 hours, so whatever error arose to lose the connection happened within the last 4 hours.

Is this a bug in node-red-contrib-google-smarthome? Or is it a glitch that's down to Google?

And is there a workaround? So is it possible to regularly test the connection and restart the flows (or just re-initialise the configuration node) if the connection test fails? I've no idea how I'd do any of that, so some pointers would be helpful. EDIT: Just found reset-server in the management node, so I guess that "re-initialise[s] the configuration node". Thanks.

**Comment by Caprico85** [comment]
Hi,

Thats interesting. I currently have similar problems with Google saying it cannot reach my home. Usually I have to wait a few minutes, then it works again, without having to restart anything. I also sometimes get the message DNS_PROBE_FINISHED_NXDOMAIN (non existant domain) when I try open my smart home website in the browser. A while later it starts to work again. Makes me wonder if Google is getting the same "non existant domain" problems as me and therefore fails to connect.

So, at first we should check if you are affected by the same problem or if it is a different reason. Therefore, I have some questions:

1. When you try to open the test page https://example.com:3001/check (or any other page on your domain) in the browser, do you also get the NXDOMAIN error from time to time, especially during the outages?

2. What DNS server are you using? For me it's Cloudflare (1.1.1.1) with fallback to Google (8.8.8.8).

3. Are you using a DynDNS provider? If so, which one? For me it's dynv6.com

4. Can you enable debug output on your management node
![image](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/2081806/fe83f83c-c69d-47c6-b181-2b403bfb6e3c)
After the next outage please check Node-RED's log file. There can be some hours difference due to timezones or DST. If a token exchange occured (successful or not) you should see a line like, along with some other debug or error messages:

   `2024-05-22T15:47:59.916359387Z POST /token 200 15.816 ms - 103`

   If you have the same problem as me, it's totally possible that you don't see any token-related messages at all.

Especially that I don't see any debug messages on my side is what makes me think that Google is having the same DNS problems and can't even connect to me.

**Comment by Caprico85** [comment]
For me, Dynv6 seemed to be the culprit. Doing "nslookup mydomain.com" (with my Dynv6 domain) regularly showed an error. On their forum are several posts that one of their nameservers is broken. I guess this is exactly the problem I had. The problem is known for at least 2 years. But they seemed to have locked down their forum and no response from the devs.

So I switched over to DuckDNS. No problems so far.
